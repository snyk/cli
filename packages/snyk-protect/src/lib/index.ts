import * as fs from 'fs';
import * as path from 'path';
import { extractPatchMetadata } from './snyk-file';
import { applyPatchToFile } from './patch';
import { findPhysicalModules } from './explore-node-modules';
import {
  VulnIdAndPackageName,
  VulnPatches,
  PatchedModule,
  ProtectResultType,
} from './types';
import { getAllPatches } from './fetch-patches';
import { sendAnalytics } from './analytics';

async function protect(projectFolderPath: string) {
  // Handle runs with flags
  if (process.argv.includes('--help')) {
    console.log(getHelp());
    return;
  }
  if (process.argv.includes('--version')) {
    console.log(getVersion());
    return;
  }

  const snykFilePath = path.resolve(projectFolderPath, '.snyk');

  if (!fs.existsSync(snykFilePath)) {
    console.log('No .snyk file found.');
    sendAnalytics({
      type: ProtectResultType.NO_SNYK_FILE,
    });
    return;
  }

  const snykFileContents = fs.readFileSync(snykFilePath, 'utf8');
  const snykFilePatchMetadata = extractPatchMetadata(snykFileContents);
  const vulnIdAndPackageNames: VulnIdAndPackageName[] = snykFilePatchMetadata;
  const targetPackageNames = [
    ...new Set(snykFilePatchMetadata.map((vpn) => vpn.packageName)), // get a list of unique package names by converting to Set and then back to array
  ];

  // find instances of the target packages by spelunking through the node_modules looking for modules with a target packageName
  const foundPhysicalPackages = findPhysicalModules(
    projectFolderPath,
    targetPackageNames,
  );

  // Map of package name to versions (for the target package names).
  // For each package name, we might have found multiple versions and we'll need to fetch patches for each version.
  // We will use this to lookup the versions of packages to get patches for.
  const packageNameToVersionsMap = new Map<string, string[]>();
  foundPhysicalPackages.forEach((p) => {
    if (packageNameToVersionsMap.has(p.packageName)) {
      const versions = packageNameToVersionsMap.get(p.packageName);
      if (!versions?.includes(p.packageVersion)) {
        versions?.push(p.packageVersion);
      }
    } else {
      packageNameToVersionsMap.set(p.packageName, [p.packageVersion]);
    }
  });

  const packageAtVersionsToPatches: Map<string, VulnPatches[]> =
    await getAllPatches(vulnIdAndPackageNames, packageNameToVersionsMap);

  if (packageAtVersionsToPatches.size === 0) {
    console.log('Nothing to patch.');
    sendAnalytics({
      type: ProtectResultType.NOTHING_TO_PATCH,
    });
    return;
  }

  const patchedModules: PatchedModule[] = [];
  foundPhysicalPackages.forEach((fpp) => {
    const packageNameAtVersion = `${fpp.packageName}@${fpp.packageVersion}`;
    const vuldIdAndPatches =
      packageAtVersionsToPatches.get(packageNameAtVersion);
    vuldIdAndPatches?.forEach((vp) => {
      vp.patches.forEach((patchDiffs) => {
        patchDiffs.patchDiffs.forEach((diff) => {
          const patchedPath = applyPatchToFile(diff, fpp.path, vp.vulnId);
          console.log(`Patched: ${patchedPath}`);
        });
      });
      patchedModules.push({
        vulnId: vp.vulnId,
        packageName: fpp.packageName,
        packageVersion: fpp.packageVersion,
      });
    });
  });

  console.log('Applied Snyk patches.');

  sendAnalytics({
    type: ProtectResultType.APPLIED_PATCHES,
    patchedModules,
  });
}

export function getHelp(): string {
  const filePath = path.resolve(__dirname, '../../help.txt');
  const helpText = fs.readFileSync(filePath, 'utf8');
  return helpText;
}

export function getVersion() {
  const filePath = path.resolve(__dirname, '../../package.json');
  const rawData = fs.readFileSync(filePath, 'utf8');
  const packageJSON = JSON.parse(rawData);
  return packageJSON.version;
}

export default protect;
