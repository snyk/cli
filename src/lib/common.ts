import * as os from 'os';
import * as alerts from './alerts';
import * as Sentry from '@sentry/node';
import * as version from './version';
import * as analytics from './analytics/index';

export async function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

export const reTryMessage =
  'Tip: Re-run in debug mode to see more information: DEBUG=*snyk* <COMMAND>';
export const contactSupportMessage =
  'If the issue persists contact support@snyk.io';

export function testPlatformSupport() {
  const supportedPlatforms = [
    'darwin amd64',
    'darwin x64',
    'darwin arm64',
    'linux amd64',
    'linux x64',
    'linux arm64',
    'win32 amd64',
    'win32 x64',
    'win32 arm64',
  ];

  const currentPlatform = os.platform() + ' ' + os.arch();
  if (!supportedPlatforms.includes(currentPlatform)) {
    const platformWarning =
      '------------------------------- Warning -------------------------------\n' +
      ' The current platform (' +
      currentPlatform +
      ') is not supported by Snyk.\n' +
      ' You may want to consider using Docker to run Snyk, for details see: https://docs.snyk.io/snyk-cli/install-the-snyk-cli#snyk-cli-in-a-docker-image\n' +
      ' If you experience errors please reach out to support@snyk.io.\n' +
      '-----------------------------------------------------------------------';

    alerts.registerAlerts([
      {
        type: 'warning',
        name: 'testPlatformSupport',
        msg: platformWarning,
      },
    ]);

    if (analytics.allowAnalytics()) {
      const sentryError = new Error('Unsupported Platform: ' + currentPlatform);
      Sentry.init({
        dsn: 'https://3e845233db8c4f43b4c4b9245f1d7bd6@o30291.ingest.sentry.io/4504599528079360',
        release: version.getVersion(),
      });
      Sentry.captureException(sentryError);
    }
  }
}
