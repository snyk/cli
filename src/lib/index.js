const snykConfig = require('./config');

// This module is kind of "world object" that is used to indirectly import modules.
// This also introduces some circular imports.

// TODO(kyegupov): untangle this, resolve circular imports, convert to Typescript

const snyk = {};
module.exports = snyk;

snyk.id = snykConfig.id;

const apiToken = require('./api-token');

// make snyk.api *always* get the latest api token from the config store
Object.defineProperty(snyk, 'api', {
  enumerable: true,
  configurable: true,
  get: function () {
    return apiToken.api();
  },
  set: function (value) {
    snykConfig.api = value;
  },
});

snyk.test = require('./snyk-test');
snyk.policy = require('snyk-policy');

// this is the user config, and not the internal config
snyk.config = require('./user-config').config;
