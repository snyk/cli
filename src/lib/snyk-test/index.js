module.exports = test;

const detect = require('../detect');
const { runTest } = require('./run-test');
const chalk = require('chalk');
const pm = require('../package-managers');
const { UnsupportedPackageManagerError } = require('../errors');
const { isMultiProjectScan } = require('../is-multi-project-scan');
const { hasFeatureFlag } = require('../feature-flags');
const {
  PNPM_FEATURE_FLAG,
  DOTNET_WITHOUT_PUBLISH_FEATURE_FLAG,
  MAVEN_DVERBOSE_EXHAUSTIVE_DEPS_FF,
} = require('../package-managers');

async function test(root, options, callback) {
  if (typeof options === 'function') {
    callback = options;
    options = {};
  }
  if (!options) {
    options = {};
  }
  const promise = executeTest(root, options);
  if (callback) {
    promise
      .then((res) => {
        callback(null, res);
      })
      .catch(callback);
  }
  return promise;
}

async function executeTest(root, options) {
  let hasPnpmSupport = false;
  let hasImprovedDotnetWithoutPublish = false;
  let enableMavenDverboseExhaustiveDeps = false;
  try {
    hasPnpmSupport = await hasFeatureFlag(PNPM_FEATURE_FLAG, options);
    if (options['dotnet-runtime-resolution']) {
      hasImprovedDotnetWithoutPublish = await hasFeatureFlag(
        DOTNET_WITHOUT_PUBLISH_FEATURE_FLAG,
        options,
      );
      if (hasImprovedDotnetWithoutPublish) {
        options.useImprovedDotnetWithoutPublish = true;
      }
    }
  } catch (err) {
    hasPnpmSupport = false;
  }

  try {
    const args = options['_doubleDashArgs'] || [];
    const verboseEnabled =
      args.includes('-Dverbose') ||
      args.includes('-Dverbose=true') ||
      !!options['print-graph'];
    if (verboseEnabled) {
      enableMavenDverboseExhaustiveDeps = await hasFeatureFlag(
        MAVEN_DVERBOSE_EXHAUSTIVE_DEPS_FF,
        options,
      );
      if (enableMavenDverboseExhaustiveDeps) {
        options.mavenVerboseIncludeAllVersions =
          enableMavenDverboseExhaustiveDeps;
      }
    }
  } catch (err) {
    enableMavenDverboseExhaustiveDeps = false;
  }

  try {
    const featureFlags = hasPnpmSupport
      ? new Set([PNPM_FEATURE_FLAG])
      : new Set([]);

    if (!options.allProjects) {
      options.packageManager = detect.detectPackageManager(
        root,
        options,
        featureFlags,
      );
    }

    return run(root, options, featureFlags).then((results) => {
      for (const res of results) {
        if (!res.packageManager) {
          res.packageManager = options.packageManager;
        }
      }
      if (results.length === 1) {
        // Return only one result if only one found as this is the default usecase
        return results[0];
      }
      // For gradle, yarnWorkspaces, allProjects we may be returning more than one result
      return results;
    });
  } catch (error) {
    return Promise.reject(
      chalk.red.bold(error.message ? error.message : error),
    );
  }
}

function run(root, options, featureFlags) {
  const projectType = options.packageManager;
  validateProjectType(options, projectType, featureFlags);
  return runTest(projectType, root, options, featureFlags);
}

function validateProjectType(options, projectType, featureFlags) {
  if (projectType === 'pnpm' && !featureFlags.has(PNPM_FEATURE_FLAG)) {
    throw new UnsupportedPackageManagerError(projectType);
  }

  if (
    !(
      options.docker ||
      isMultiProjectScan(options) ||
      pm.SUPPORTED_PACKAGE_MANAGER_NAME[projectType]
    )
  ) {
    throw new UnsupportedPackageManagerError(projectType);
  }
}
