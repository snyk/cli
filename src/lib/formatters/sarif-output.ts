import * as sarif from 'sarif';
import * as upperFirst from 'lodash.upperfirst';
import { AnnotatedIssue, TestResult } from '../snyk-test/legacy';
import { SEVERITY } from '../snyk-test/legacy';
import { getResults } from './get-sarif-result';
import { getVersion } from '../version';

export function createSarifOutputForContainers(
  testResults: TestResult[],
): sarif.Log {
  const sarifRes: sarif.Log = {
    $schema:
      'https://raw.githubusercontent.com/oasis-tcs/sarif-spec/master/Schemata/sarif-schema-2.1.0.json',
    version: '2.1.0',
    runs: [],
  };

  testResults.forEach((testResult) => {
    sarifRes.runs.push({
      automationDetails : getAutomationDetails(testResult),
      tool: getTool(testResult),
      results: getResults(testResult),
    });
  });

  return sarifRes;
}

export function getIssueLevel(
  severity: SEVERITY | 'none',
): sarif.ReportingConfiguration.level {
  return severity === SEVERITY.HIGH || severity === SEVERITY.CRITICAL
    ? 'error'
    : 'warning';
}

export function getTool(testResult): sarif.Tool {
  const tool: sarif.Tool = {
    driver: {
      name: 'Snyk Container',
      semanticVersion: getVersion(),
      version: getVersion(),
      informationUri: 'https://docs.snyk.io/',
      properties: {
        artifactsScanned: testResult.dependencyCount,
      },
      rules: [],
    },
  };

  if (!testResult.vulnerabilities) {
    return tool;
  }

  const pushedIds = {};
  tool.driver.rules = testResult.vulnerabilities
    .map((vuln: AnnotatedIssue) => {
      if (pushedIds[vuln.id]) {
        return;
      }
      const level = getIssueLevel(vuln.severity);
      const cve = vuln.identifiers?.CVE?.join();
      pushedIds[vuln.id] = true;
      return {
        id: vuln.id,
        shortDescription: {
          text: `${upperFirst(vuln.severity)} severity - ${
            vuln.title
          } vulnerability in ${vuln.packageName}`,
        },
        fullDescription: {
          text: cve
            ? `(${cve}) ${vuln.name}@${vuln.version}`
            : `${vuln.name}@${vuln.version}`,
        },
        help: {
          text: '',
          markdown: vuln.description,
        },
        defaultConfiguration: {
          level: level,
        },
        properties: {
          tags: [
            'security',
            ...(vuln.identifiers?.CWE || []),
            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
            testResult.packageManager!,
          ],
          cvssv3_baseScore: vuln.cvssScore, // AWS
          'security-severity': String(vuln.cvssScore), // GitHub
        },
      };
    })
    .filter(Boolean);
  return tool;
}

// Github anncouned changes to their SARIF upload -- https://github.blog/changelog/2024-05-06-code-scanning-will-stop-combining-runs-from-a-single-upload/
// the impact is when a SARIF that is being uploaded, each run must have unique category, as defined by GitHub here, https://docs.github.com/en/code-security/code-scanning/integrating-with-code-scanning/sarif-support-for-code-scanning#runautomationdetails-object
// This presents a new problem of when a file is removed from source since GH will not have an empty result to close any previously opened items since GH open/closes
// based on the SARIF tool.driver.name + Category. Open source's solution is the most obvious, inlcude the targetFile. Snyk-iac, is using this field set to a static "snyk-iac". Combing what
// was being done there with the file name to generate the unique value. Using | as a separator to make it easier to parse out tool vs targetFile. 
function getAutomationDetails(testResult: TestResult)
{
  let automationId = !!process.env.SET_AUTOMATION_DETAILS_ID ? `snyk-container|${testResult.displayTargetFile || testResult.targetFile }/` : ""
  return {
    id : automationId,
  };
}