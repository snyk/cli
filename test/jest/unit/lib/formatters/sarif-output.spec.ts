import { createSarifOutputForContainers } from '../../../../../src/lib/formatters/sarif-output';
import { SEVERITY, TestResult } from '../../../../../src/lib/snyk-test/legacy';
import { SupportedProjectTypes } from '../../../../../src/lib/types';

jest.useFakeTimers().setSystemTime(new Date('2025-01-01T00:00:00.000Z'));

describe('createSarifOutputForContainers', () => {
  it('general with high severity issue', () => {
    const identifiers = {
      ALTERNATIVE: [
        'SNYK-DEBIAN8-EXPAT-450909',
        'SNYK-DEBIAN9-EXPAT-450910',
        'SNYK-DEBIANUNSTABLE-EXPAT-450911',
        'SNYK-DEBIAN10-EXPAT-450912',
        'SNYK-UBUNTU1404-EXPAT-451027',
        'SNYK-UBUNTU1204-EXPAT-451028',
        'SNYK-UBUNTU1810-EXPAT-451029',
        'SNYK-UBUNTU1604-EXPAT-451030',
        'SNYK-UBUNTU1804-EXPAT-451031',
        'SNYK-ALPINE38-EXPAT-451858',
        'SNYK-ALPINE39-EXPAT-453353',
        'SNYK-ALPINE37-EXPAT-453374',
        'SNYK-ALPINE310-EXPAT-453902',
        'SNYK-DEBIAN11-EXPAT-518187',
        'SNYK-UBUNTU1904-EXPAT-531483',
      ],
      CVE: ['CVE-2018-20843'],
      CWE: ['CWE-611'],
    };
    const testFile = getTestResult(SEVERITY.HIGH, identifiers);
    const sarif = createSarifOutputForContainers([testFile]);
    expect(sarif).toMatchSnapshot();
  });

  it('general with critical severity issue without CVE', () => {
    const identifiersWithoutCVE = {
      ALTERNATIVE: [
        'SNYK-DEBIAN8-EXPAT-450909',
        'SNYK-DEBIAN9-EXPAT-450910',
        'SNYK-DEBIANUNSTABLE-EXPAT-450911',
        'SNYK-DEBIAN10-EXPAT-450912',
        'SNYK-UBUNTU1404-EXPAT-451027',
        'SNYK-UBUNTU1204-EXPAT-451028',
        'SNYK-UBUNTU1810-EXPAT-451029',
        'SNYK-UBUNTU1604-EXPAT-451030',
        'SNYK-UBUNTU1804-EXPAT-451031',
        'SNYK-ALPINE38-EXPAT-451858',
        'SNYK-ALPINE39-EXPAT-453353',
        'SNYK-ALPINE37-EXPAT-453374',
        'SNYK-ALPINE310-EXPAT-453902',
        'SNYK-DEBIAN11-EXPAT-518187',
        'SNYK-UBUNTU1904-EXPAT-531483',
      ],
      CWE: ['CWE-611'],
    };
    const testFile = getTestResult(SEVERITY.CRITICAL, identifiersWithoutCVE);
    const sarif = createSarifOutputForContainers([testFile]);
    expect(sarif).toMatchSnapshot();
  });

  it('general with critical severity issue without CWE', () => {
    const identifiersWithoutCWE = {
      ALTERNATIVE: [
        'SNYK-DEBIAN8-EXPAT-450909',
        'SNYK-DEBIAN9-EXPAT-450910',
        'SNYK-DEBIANUNSTABLE-EXPAT-450911',
        'SNYK-DEBIAN10-EXPAT-450912',
        'SNYK-UBUNTU1404-EXPAT-451027',
        'SNYK-UBUNTU1204-EXPAT-451028',
        'SNYK-UBUNTU1810-EXPAT-451029',
        'SNYK-UBUNTU1604-EXPAT-451030',
        'SNYK-UBUNTU1804-EXPAT-451031',
        'SNYK-ALPINE38-EXPAT-451858',
        'SNYK-ALPINE39-EXPAT-453353',
        'SNYK-ALPINE37-EXPAT-453374',
        'SNYK-ALPINE310-EXPAT-453902',
        'SNYK-DEBIAN11-EXPAT-518187',
        'SNYK-UBUNTU1904-EXPAT-531483',
      ],
      CVE: ['CVE-2018-20843'],
    };
    const testFile = getTestResult(SEVERITY.CRITICAL, identifiersWithoutCWE);
    const sarif = createSarifOutputForContainers([testFile]);
    expect(sarif).toMatchSnapshot();
  });
});

function getTestResult(
  severity: SEVERITY,
  identifiers?: {
    [name: string]: string[];
  },
): TestResult {
  return {
    vulnerabilities: [
      {
        credit: [''],
        description:
          '## Overview\nIn libexpat in Expat before 2.2.7, XML input including XML names that contain a large number of colons could make the XML parser consume a high amount of RAM and CPU resources while processing (enough to be usable for denial-of-service attacks).\n\n## References\n- [Bugtraq Mailing List](https://seclists.org/bugtraq/2019/Jun/39)\n- [CVE Details](https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2018-20843)\n- [Debian Security Advisory](https://www.debian.org/security/2019/dsa-4472)\n- [Debian Security Announcement](https://lists.debian.org/debian-lts-announce/2019/06/msg00028.html)\n- [Debian Security Tracker](https://security-tracker.debian.org/tracker/CVE-2018-20843)\n- [GitHub Commit](https://github.com/libexpat/libexpat/pull/262/commits/11f8838bf99ea0a6f0b76f9760c43704d00c4ff6)\n- [GitHub Issue](https://github.com/libexpat/libexpat/issues/186)\n- [GitHub PR](https://github.com/libexpat/libexpat/pull/262)\n- [MISC](https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=5226)\n- [MISC](https://github.com/libexpat/libexpat/blob/R_2_2_7/expat/Changes)\n- [Netapp Security Advisory](https://security.netapp.com/advisory/ntap-20190703-0001/)\n- [Ubuntu CVE Tracker](http://people.ubuntu.com/~ubuntu-security/cve/CVE-2018-20843)\n- [Ubuntu Security Advisory](https://usn.ubuntu.com/4040-1/)\n- [Ubuntu Security Advisory](https://usn.ubuntu.com/4040-2/)\n',
        id: 'SNYK-LINUX-EXPAT-450908',
        cvssScore: 7.5,
        identifiers,
        packageManager: 'linux' as SupportedProjectTypes,
        packageName: 'expat',
        patches: [],
        publicationTime: '2019-06-24T22:21:12.802637Z',
        semver: {
          vulnerableByDistro: {
            'alpine:3.10': ['<2.2.7-r0'],
            'alpine:3.7': ['<2.2.7-r0'],
            'alpine:3.8': ['<2.2.7-r0'],
            'alpine:3.9': ['<2.2.7-r0'],
            'debian:10': ['<2.2.6-2'],
            'debian:11': ['<2.2.6-2'],
            'debian:8': ['<2.1.0-6+deb8u5'],
            'debian:9': ['<2.2.0-2+deb9u2'],
            'debian:unstable': ['<2.2.6-2'],
            'ubuntu:12.04': ['<2.0.1-7.2ubuntu1.6'],
            'ubuntu:14.04': ['<2.1.0-4ubuntu1.4+esm1'],
            'ubuntu:16.04': ['<2.1.0-7ubuntu0.16.04.4'],
            'ubuntu:18.04': ['<2.2.5-3ubuntu0.1'],
            'ubuntu:18.10': ['<2.2.6-1ubuntu0.18.10'],
            'ubuntu:19.04': ['<2.2.6-1ubuntu0.19.04'],
          },
          vulnerable: ['<2.2.7-r0'],
        },
        severity,
        title: 'XML External Entity (XXE) Injection',
        from: [
          'docker-image|garethr/snyky@alpine',
          '.python-rundeps@0',
          'expat@2.2.5-r0',
        ],
        upgradePath: [],
        isUpgradable: false,
        isPatchable: false,
        name: 'expat',
        version: '2.2.5-r0',
        fixedIn: ['2.2.7-r0'],
        parentDepType: 'prod',
        isNew: false,
        below: '',
      },
    ],
    ok: false,
    dependencyCount: 969,
    org: 'ORG',
    policy: '',
    isPrivate: true,
    licensesPolicy: {
      severities: {},
      orgLicenseRules: {},
    },
    packageManager: 'npm',
    ignoreSettings: null,
    summary: '165 vulnerable dependencies',
    filesystemPolicy: false,
    uniqueCount: 42,
    projectName: 'PROJECT_NAME',
    foundProjectCount: 22,
    displayTargetFile: 'Dockerfile',
  };
}
