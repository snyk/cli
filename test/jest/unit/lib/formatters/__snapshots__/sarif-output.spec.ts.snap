// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createSarifOutputForContainers general with critical severity issue without CVE 1`] = `
{
  "$schema": "https://raw.githubusercontent.com/oasis-tcs/sarif-spec/master/Schemata/sarif-schema-2.1.0.json",
  "runs": [
    {
      "automationDetails": {
        "id": "",
      },
      "results": [
        {
          "fixes": undefined,
          "level": "error",
          "locations": [
            {
              "logicalLocations": [
                {
                  "fullyQualifiedName": "expat@2.2.5-r0",
                },
              ],
              "physicalLocation": {
                "artifactLocation": {
                  "uri": "Dockerfile",
                },
                "region": {
                  "startLine": 1,
                },
              },
            },
          ],
          "message": {
            "text": "This file introduces a vulnerable expat package with a critical severity vulnerability.",
          },
          "ruleId": "SNYK-LINUX-EXPAT-450908",
        },
      ],
      "tool": {
        "driver": {
          "informationUri": "https://docs.snyk.io/",
          "name": "Snyk Container",
          "properties": {
            "artifactsScanned": 969,
          },
          "rules": [
            {
              "defaultConfiguration": {
                "level": "error",
              },
              "fullDescription": {
                "text": "expat@2.2.5-r0",
              },
              "help": {
                "markdown": "## Overview
In libexpat in Expat before 2.2.7, XML input including XML names that contain a large number of colons could make the XML parser consume a high amount of RAM and CPU resources while processing (enough to be usable for denial-of-service attacks).

## References
- [Bugtraq Mailing List](https://seclists.org/bugtraq/2019/Jun/39)
- [CVE Details](https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2018-20843)
- [Debian Security Advisory](https://www.debian.org/security/2019/dsa-4472)
- [Debian Security Announcement](https://lists.debian.org/debian-lts-announce/2019/06/msg00028.html)
- [Debian Security Tracker](https://security-tracker.debian.org/tracker/CVE-2018-20843)
- [GitHub Commit](https://github.com/libexpat/libexpat/pull/262/commits/11f8838bf99ea0a6f0b76f9760c43704d00c4ff6)
- [GitHub Issue](https://github.com/libexpat/libexpat/issues/186)
- [GitHub PR](https://github.com/libexpat/libexpat/pull/262)
- [MISC](https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=5226)
- [MISC](https://github.com/libexpat/libexpat/blob/R_2_2_7/expat/Changes)
- [Netapp Security Advisory](https://security.netapp.com/advisory/ntap-20190703-0001/)
- [Ubuntu CVE Tracker](http://people.ubuntu.com/~ubuntu-security/cve/CVE-2018-20843)
- [Ubuntu Security Advisory](https://usn.ubuntu.com/4040-1/)
- [Ubuntu Security Advisory](https://usn.ubuntu.com/4040-2/)
",
                "text": "",
              },
              "id": "SNYK-LINUX-EXPAT-450908",
              "properties": {
                "cvssv3_baseScore": 7.5,
                "security-severity": "7.5",
                "tags": [
                  "security",
                  "CWE-611",
                  "npm",
                ],
              },
              "shortDescription": {
                "text": "Critical severity - XML External Entity (XXE) Injection vulnerability in expat",
              },
            },
          ],
          "semanticVersion": "1.0.0-monorepo",
          "version": "1.0.0-monorepo",
        },
      },
    },
  ],
  "version": "2.1.0",
}
`;

exports[`createSarifOutputForContainers general with critical severity issue without CWE 1`] = `
{
  "$schema": "https://raw.githubusercontent.com/oasis-tcs/sarif-spec/master/Schemata/sarif-schema-2.1.0.json",
  "runs": [
    {
      "automationDetails": {
        "id": "",
      },
      "results": [
        {
          "fixes": undefined,
          "level": "error",
          "locations": [
            {
              "logicalLocations": [
                {
                  "fullyQualifiedName": "expat@2.2.5-r0",
                },
              ],
              "physicalLocation": {
                "artifactLocation": {
                  "uri": "Dockerfile",
                },
                "region": {
                  "startLine": 1,
                },
              },
            },
          ],
          "message": {
            "text": "This file introduces a vulnerable expat package with a critical severity vulnerability.",
          },
          "ruleId": "SNYK-LINUX-EXPAT-450908",
        },
      ],
      "tool": {
        "driver": {
          "informationUri": "https://docs.snyk.io/",
          "name": "Snyk Container",
          "properties": {
            "artifactsScanned": 969,
          },
          "rules": [
            {
              "defaultConfiguration": {
                "level": "error",
              },
              "fullDescription": {
                "text": "(CVE-2018-20843) expat@2.2.5-r0",
              },
              "help": {
                "markdown": "## Overview
In libexpat in Expat before 2.2.7, XML input including XML names that contain a large number of colons could make the XML parser consume a high amount of RAM and CPU resources while processing (enough to be usable for denial-of-service attacks).

## References
- [Bugtraq Mailing List](https://seclists.org/bugtraq/2019/Jun/39)
- [CVE Details](https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2018-20843)
- [Debian Security Advisory](https://www.debian.org/security/2019/dsa-4472)
- [Debian Security Announcement](https://lists.debian.org/debian-lts-announce/2019/06/msg00028.html)
- [Debian Security Tracker](https://security-tracker.debian.org/tracker/CVE-2018-20843)
- [GitHub Commit](https://github.com/libexpat/libexpat/pull/262/commits/11f8838bf99ea0a6f0b76f9760c43704d00c4ff6)
- [GitHub Issue](https://github.com/libexpat/libexpat/issues/186)
- [GitHub PR](https://github.com/libexpat/libexpat/pull/262)
- [MISC](https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=5226)
- [MISC](https://github.com/libexpat/libexpat/blob/R_2_2_7/expat/Changes)
- [Netapp Security Advisory](https://security.netapp.com/advisory/ntap-20190703-0001/)
- [Ubuntu CVE Tracker](http://people.ubuntu.com/~ubuntu-security/cve/CVE-2018-20843)
- [Ubuntu Security Advisory](https://usn.ubuntu.com/4040-1/)
- [Ubuntu Security Advisory](https://usn.ubuntu.com/4040-2/)
",
                "text": "",
              },
              "id": "SNYK-LINUX-EXPAT-450908",
              "properties": {
                "cvssv3_baseScore": 7.5,
                "security-severity": "7.5",
                "tags": [
                  "security",
                  "npm",
                ],
              },
              "shortDescription": {
                "text": "Critical severity - XML External Entity (XXE) Injection vulnerability in expat",
              },
            },
          ],
          "semanticVersion": "1.0.0-monorepo",
          "version": "1.0.0-monorepo",
        },
      },
    },
  ],
  "version": "2.1.0",
}
`;

exports[`createSarifOutputForContainers general with high severity issue 1`] = `
{
  "$schema": "https://raw.githubusercontent.com/oasis-tcs/sarif-spec/master/Schemata/sarif-schema-2.1.0.json",
  "runs": [
    {
      "automationDetails": {
        "id": "",
      },
      "results": [
        {
          "fixes": undefined,
          "level": "error",
          "locations": [
            {
              "logicalLocations": [
                {
                  "fullyQualifiedName": "expat@2.2.5-r0",
                },
              ],
              "physicalLocation": {
                "artifactLocation": {
                  "uri": "Dockerfile",
                },
                "region": {
                  "startLine": 1,
                },
              },
            },
          ],
          "message": {
            "text": "This file introduces a vulnerable expat package with a high severity vulnerability.",
          },
          "ruleId": "SNYK-LINUX-EXPAT-450908",
        },
      ],
      "tool": {
        "driver": {
          "informationUri": "https://docs.snyk.io/",
          "name": "Snyk Container",
          "properties": {
            "artifactsScanned": 969,
          },
          "rules": [
            {
              "defaultConfiguration": {
                "level": "error",
              },
              "fullDescription": {
                "text": "(CVE-2018-20843) expat@2.2.5-r0",
              },
              "help": {
                "markdown": "## Overview
In libexpat in Expat before 2.2.7, XML input including XML names that contain a large number of colons could make the XML parser consume a high amount of RAM and CPU resources while processing (enough to be usable for denial-of-service attacks).

## References
- [Bugtraq Mailing List](https://seclists.org/bugtraq/2019/Jun/39)
- [CVE Details](https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2018-20843)
- [Debian Security Advisory](https://www.debian.org/security/2019/dsa-4472)
- [Debian Security Announcement](https://lists.debian.org/debian-lts-announce/2019/06/msg00028.html)
- [Debian Security Tracker](https://security-tracker.debian.org/tracker/CVE-2018-20843)
- [GitHub Commit](https://github.com/libexpat/libexpat/pull/262/commits/11f8838bf99ea0a6f0b76f9760c43704d00c4ff6)
- [GitHub Issue](https://github.com/libexpat/libexpat/issues/186)
- [GitHub PR](https://github.com/libexpat/libexpat/pull/262)
- [MISC](https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=5226)
- [MISC](https://github.com/libexpat/libexpat/blob/R_2_2_7/expat/Changes)
- [Netapp Security Advisory](https://security.netapp.com/advisory/ntap-20190703-0001/)
- [Ubuntu CVE Tracker](http://people.ubuntu.com/~ubuntu-security/cve/CVE-2018-20843)
- [Ubuntu Security Advisory](https://usn.ubuntu.com/4040-1/)
- [Ubuntu Security Advisory](https://usn.ubuntu.com/4040-2/)
",
                "text": "",
              },
              "id": "SNYK-LINUX-EXPAT-450908",
              "properties": {
                "cvssv3_baseScore": 7.5,
                "security-severity": "7.5",
                "tags": [
                  "security",
                  "CWE-611",
                  "npm",
                ],
              },
              "shortDescription": {
                "text": "High severity - XML External Entity (XXE) Injection vulnerability in expat",
              },
            },
          ],
          "semanticVersion": "1.0.0-monorepo",
          "version": "1.0.0-monorepo",
        },
      },
    },
  ],
  "version": "2.1.0",
}
`;
